pkg/volume/csi/csi_plugin.go

```diff
@@ -197,15 +202,62 @@ func (h *RegistrationHandler) validateVersions(callerName, pluginName string, en
 	return newDriverHighestVersion, nil
 }
 
+// newCsiDriverClient creates a CSI driver client based on a driver name.
+func (p *csiPlugin) newCsiDriverClient(driverName csiDriverName) (*csiDriverClient, error) {
+	if driverName == "" {
+		return nil, fmt.Errorf("driver name is empty")
+	}
+
+	addr := fmt.Sprintf(csiAddrTemplate, driverName)
+	requiresV0Client := true
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
+		existingDriver, driverExists := p.drivers.Get(string(driverName))
+		if !driverExists {
+			return nil, fmt.Errorf("driver name %s not found in the list of registered CSI drivers", driverName)
+		}
+
+		addr = existingDriver.endpoint
+		requiresV0Client = versionRequiresV0Client(existingDriver.highestSupportedVersion)
+	}
+
+	nodeV1ClientCreator := newV1NodeClient
+	nodeV0ClientCreator := newV0NodeClient
+	if requiresV0Client {
+		nodeV1ClientCreator = nil
+	} else {
+		nodeV0ClientCreator = nil
+	}
+
+	return &csiDriverClient{
+		driverName:          driverName,
+		addr:                csiAddr(addr),
+		nodeV1ClientCreator: nodeV1ClientCreator,
+		nodeV0ClientCreator: nodeV0ClientCreator,
+	}, nil
+}
+
+// getClientCreator takes a driver name and returns a function that constructs
+// a new CSI driver client of that name/type.
+func (p *csiPlugin) getClientCreator(driverName csiDriverName) func() (*csiDriverClient, error) {
```

Comment:
Do we still need this level of indirection when creating the client.  Why not use `csi_plugin.newCsiDriverClient` directly?

Reply:
This func is there, to help a bit with the `csiClientGetter`, e.g. like here: https://github.com/kubernetes/kubernetes/pull/74966/files#diff-f52821ec6236e86b02ca334533395c91R523.

Would you prefer to store both
- the driver name
- the `newCsiDriverClient` func
on the `csiClientGetter` struct and call the latter with the former on a `csiClientGetter.Get()` call?

Happy to do so. TBH it feels a bit nicer to me as it is, but I have no strong opinions about that.

Reply:
We can see if we can streamline the lazy creation and/or caching of the client in a separate PR later, we might be able to combine different aspects of `csiPlugin`, `csiClientGetter`, `getClientCreator`, and `newCsiDriverClient` -- and make the code easier to understand by doing so.